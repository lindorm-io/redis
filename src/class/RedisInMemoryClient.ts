import { IRedisClientOptions } from "./RedisClient";
import { TObject } from "@lindorm-io/core";
import { includes } from "lodash";
import { parseBlob, stringifyBlob } from "../util";

export class RedisInMemoryClient {
  public store: TObject<any>;

  constructor(options: IRedisClientOptions) {
    this.store = {};
  }

  public async set(key: string, value: TObject<any>, expiresInSeconds?: number): Promise<string> {
    this.store[key] = { blob: stringifyBlob(value), expiresInSeconds };

    return "OK";
  }

  public async get(key: string): Promise<TObject<any>> {
    const data = this.store[key];

    return data?.blob ? parseBlob(data.blob) : undefined;
  }

  public async getAll(pattern: string): Promise<Array<TObject<any>>> {
    const array: Array<TObject<any>> = [];
    const inMemoryPattern = pattern.replace(/\*/g, "");

    for (const key of Object.keys(this.store)) {
      if (!includes(key, inMemoryPattern)) continue;
      array.push(await this.get(key));
    }

    return array;
  }

  public async del(key: string): Promise<number> {
    this.store[key] = undefined;
    return 1;
  }
}
