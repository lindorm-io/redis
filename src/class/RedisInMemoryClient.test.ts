import { RedisInMemoryClient } from "./RedisInMemoryClient";

describe("RedisInMemoryClient", () => {
  let client: RedisInMemoryClient;

  beforeEach(async () => {
    client = new RedisInMemoryClient();
    await client.connect();
  });

  test("should quit", async () => {
    await expect(client.quit()).resolves.toBe("OK");
  });

  test("should set", async () => {
    await expect(client.set("key", { blob: { object: { is: true } } })).resolves.toBe("OK");
    expect(client.cache).toMatchSnapshot();
  });

  test("should get", async () => {
    await expect(client.set("key", { blob: "yes" })).resolves.toBe("OK");

    await expect(client.get("key")).resolves.toMatchSnapshot();
  });

  test("should get all", async () => {
    await expect(client.set("key1", { blob: "yes" })).resolves.toBe("OK");
    await expect(client.set("key2", { blob: "no" })).resolves.toBe("OK");

    await expect(client.getAll("key*")).resolves.toMatchSnapshot();
  });

  test("should delete", async () => {
    await expect(client.set("key", { blob: "yes" })).resolves.toBe("OK");

    await expect(client.del("key")).resolves.toBe(1);
    expect(client.cache).toMatchSnapshot();
  });
});
