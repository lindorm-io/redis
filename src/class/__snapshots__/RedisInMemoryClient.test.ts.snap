// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedisInMemoryClient should delete 1`] = `
Object {
  "key": undefined,
}
`;

exports[`RedisInMemoryClient should get 1`] = `
Object {
  "blob": "yes",
}
`;

exports[`RedisInMemoryClient should get all 1`] = `
Array [
  Object {
    "blob": "yes",
  },
  Object {
    "blob": "no",
  },
]
`;

exports[`RedisInMemoryClient should set 1`] = `
Object {
  "key": Object {
    "blob": "{\\"json\\":{\\"blob\\":{\\"object\\":{\\"is\\":\\"true\\"}}},\\"meta\\":{\\"blob\\":{\\"object\\":{\\"is\\":\\"boolean\\"}}}}",
    "expiresInSeconds": undefined,
  },
}
`;
